### DISCUSSION TITLE
Supporting EdDSA

### DISCUSSION THREAD

**Discussion by tnorimat** [keycloak/keycloak#15713] [Ideas]
Hello.

I would like to propose supporting EdDSA digital signature algorithm for keycloak.
I created the corresponding [issue](https://github.com/keycloak/keycloak/issues/15714) describing its specification in detail.

### Motivation
 Keycloak can be widely acceptable if it supports security standards.

Some standards defined EdDSA.
  - [IRTF RFC 7748 Elliptic Curves for Security](https://datatracker.ietf.org/doc/html/rfc7748)
    It picked up Curve25519 and Curve448 as recommended curves.
  - [IRTF RFC 8032 Edwards-Curve Digital Signature Algorithm (EdDSA)](https://datatracker.ietf.org/doc/html/rfc8032)
    It defined EdDSA by providing specific parameters of Curve25519 and Curve448.
    Curve25519: Ed25519ph, Ed25519ctx, and Ed25519
    Curve448: Ed448ph and Ed448
  - [IRTF RFC 8037 CFRG Elliptic Curve Diffie Hellman (ECDH) and Signatures in JSON Object Signing and Encryption (JOSE)](https://www.rfc-editor.org/rfc/rfc8037.html)
    It defined how to use EdDSA algorithms (Ed25519 and Ed448) in JOSE in an interoperable manner.
  - [NIST FIPS PUB 186-5 (Draft) Digital Signature Standard (DSS)](https://csrc.nist.gov/publications/detail/fips/186/5/draft)
    It refered RFC 8032 and only picked up Ed25519ph, Ed25519, Ed448ph and Ed448 (excluding Ed25519ctx).
  - [NIST SP 800-186 (Draft) Recommendations for Discrete Logarithm-Based Cryptography: Elliptic Curve Domain Parameters](https://csrc.nist.gov/publications/detail/sp/800-186/draft)
    It defined Edwards-curves defined in RFC 7748 and RFC 8032.

  Other standards that are not directly related to keycloak adopted EdDSA.
  - TLS
    [RFC 8446 The Transport Layer Security (TLS) Protocol Version 1.3](https://www.rfc-editor.org/rfc/rfc8446)
  - FIDO
    [FIDO Authenticator Allowed Cryptography List](https://fidoalliance.org/specs/fido-security-requirements-v1.0-fd-20170524/fido-authenticator-allowed-cryptography-list_20170524.html)
  - WebAuthn
    [Web Authentication: An API for accessing Public Key Credentials Level 2](https://www.w3.org/TR/webauthn-2/)
  - Java Cryptography Architecture (JCA)
    [JEP 339: Edwards-Curve Digital Signature Algorithm (EdDSA)](https://openjdk.org/jeps/339)
    It implemented RFC 8032 and was supported by JDK 15.

 Other OSS supported EdDSA.
  - JDK (from version 15 mentioned above)
  - [Bouncy Castle](https://www.bouncycastle.org/specifications.html)
    It supported Ed25519 and Ed448.
  - [OpenSSL](https://www.openssl.org/docs/man1.1.1/man7/Ed25519.html)
    It supported Ed25519 and Ed448.
  - [OpenSSH](https://www.openssh.com/specs.html)
  - [Golang](https://pkg.go.dev/github.com/katzenpost/core/crypto/eddsa)
  - [Libgcrypt](https://www.gnupg.org/documentation/manuals/gcrypt/Cryptographic-Functions.html)

 As for security of EdDSA, it was examined in some reports. The following report concluded that EdDSA using Curve25519 is a
good signature scheme for use up to 2030.
  - [CRYPTREC Review of EdDSA](https://www.cryptrec.go.jp/exreport/cryptrec-ex-3003-2020.pdf)

 As for performance of EdDSA (Ed25519 is considered to have 128 security bits) compared with ECDSA (P-256 is considered to have 128 security bits), it was examed in some benchmark reports. According to the following benchmark report, it might be ok to say that Ed25519 takes less CPU cycles compared withh ECDSA P-256 for key generation, signing and verifying a sign.
  - [eBACS: ECRYPT Benchmarking of Cryptographic Systems](https://bench.cr.yp.to/results-sign.html)

 Considering them, it is beneficial to support EdDSA for keycloak.

**Comment by codespearhead** [comment]
It seems that issue #15714 was closed. Should this discussion be closed as well?

**Comment by tnorimat** [comment]
@codespearhead Thank you. I close the discussion.
